File: 1255. Maximum Score Words Formed by Letters/maxScoreWords.js
/**
 * @param {string[]} words
 * @param {character[]} letters
 * @param {number[]} score
 * @return {number}
 */
const maxScoreWords = (words, letters, score) => {
  const lettersCnt = {};
  for (let i = 0; i < letters.length; i++) {
    const c = letters[i];
    if (!lettersCnt[c]) {
      lettersCnt[c] = 0;
    }
    lettersCnt[c]++;
  }

  const canFormWord = (word, lettersCnt) => {
    const wordCnt = {};
    for (let i = 0; i < word.length; i++) {
      const c = word[i];
      if (!wordCnt[c]) {
        wordCnt[c] = 0;
      }
      wordCnt[c]++;
    }

    for (let j = 0; j < word.length; j++) {
      const c = word[j];
      if (wordCnt[c] > (lettersCnt[c] || 0)) {
        return false;
      }
    }

    return true;
  };

  const getScore = (word) => {
    let res = 0;
    let charA = "a".charCodeAt(0);
    for (let i = 0; i < word.length; i++) {
      const c = word[i]
      const charCode = c.charCodeAt(0);
      const index = charCode - charA;
      
      res += score[index];
    }

    return res;
  }

  const backtracking = (i) => {
    if (i === words.length) {
      return 0;
    }
    // skip the word
    let res = backtracking(i + 1);

    if (canFormWord(words[i], lettersCnt)) {
      for (let j = 0; j < words[i].length; j++) {
        const c = words[i][j];
        lettersCnt[c]--;
      }

      res = Math.max(res, getScore(words[i]) + backtracking(i + 1));

      for (let j = 0; j < words[i].length; j++) {
        const c = words[i][j];
        lettersCnt[c]++;
      }
    }
    
    return res;
  }

  return backtracking(0);
};


const words = ["dog", "cat", "dad", "good"];
const letters = ["a", "a", "c", "d", "d", "d", "g", "o", "o"];
const score = [1, 0, 9, 5, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];

console.log('maxScoreWords: ', maxScoreWords(words, letters, score))

File: 1255. Maximum Score Words Formed by Letters/maxScoreWords.py
from collections import Counter
from typing import List

class Solution:
  def maxScoreWords(self, words: List[str], letters: List[str], score: List[int]) -> int:
    def can_form_word(word, letter_cnt):
      word_cnt = Counter(word)
      for c in word_cnt:
        if word_cnt[c] > letter_cnt[c]:
          return False
      return True
    
    def get_score(word):
      res = 0
      for c in word:
        res += score[ord(c) - ord('a')]
      return res
      
    letter_cnt = Counter(letters)
    
    def backtracking(i):
      if i == len(words):
        return 0
      # skip the word
      res = backtracking(i + 1)
      
      if can_form_word(words[i], letter_cnt):
        for c in words[i]:
          letter_cnt[c] -= 1
        res = max(res, get_score(words[i]) + backtracking(i + 1))
        for c in words[i]:
          letter_cnt[c] += 1
      return res
    
    return backtracking(0)
  

words = ["dog","cat","dad","good"]
letters = ["a","a","c","d","d","d","g","o","o"]
score = [1,0,9,5,0,0,3,0,0,0,0,0,0,0,2,0,0,0,0,0,0,0,0,0,0,0]

solution = Solution()
print('maxScoreWords: ', solution.maxScoreWords(words, letters, score))
